weave/trace_server/calls_query_builder.py

```diff
@@ -0,0 +1,726 @@
+"""
+This module builds on the orm.py module to provide a more hard-coded optimized
+query builder specifically for the "Calls" table (which is the `calls_merged`
+underlying table).
+
+The `CallsQuery` class is the main entry point for building a query. These
+optimizations are hand-tuned for Clickhouse - and in particular attempt to
+perform predicate pushdown where possible and delay loading expensive fields as
+much as possible. In testing, Clickhouse performance is dominated by the amount
+of data loaded into memory.
+
+Outstanding Optimizations/Work:
+
+* [ ] The CallsQuery API itself is a little clunky with the returning self pattern. Consider revision
+* [ ] This code could use more of the orm.py code to reduce logical duplication,
+  specifically:
+
+        1. `process_query_to_conditions` is nearly identical to the one in
+        orm.py, but differs enough that generalizing a common function was not
+        trivial.
+        2. We define our own column definitions here, which might be
+        able to use the ones in orm.py.
+
+* [ ] Implement column selection from callers
```

Comment:
Not sure if this meant "selection by callers" or "selection from calls"

Reply:
elaborated